sklearn/neural_network/tests/test_mlp.py

```diff
@@ -876,3 +892,14 @@ def test_mlp_loading_from_joblib_partial_fit(tmp_path):
     # finetuned model learned the new target
     predicted_value = load_estimator.predict(fine_tune_features)
     assert_allclose(predicted_value, fine_tune_target, rtol=1e-4)
+
+
+@pytest.mark.parametrize("MLPEstimator", [MLPClassifier, MLPRegressor])
+def test_mlp_warm_start_with_early_stopping(MLPEstimator):
+    """Check that early stopping works with warm start."""
+    mlp = MLPEstimator(
+        max_iter=10, random_state=0, warm_start=True, early_stopping=True
+    )
+    mlp.fit(X_iris, y_iris)
+    mlp.set_params(max_iter=20)
+    mlp.fit(X_iris, y_iris)
```

Comment:
Can we check that length of `validation_scores_` to make sure it is updated?

Reply:
I think that there is an underlying bug: with `max_iter=20`, we end up with `n_iter_ == 30` for the regressor. I need to investigate. But it could be fixed in another PR.

Reply:
Isn't this because there were ten (original) + twenty (warm start) iterations?

Reply:
Not sure yet. I have to go into details. I have an issue understanding the semantics there.

Having `n_iter_ > max_iter` is surprising. Calling `fit` just mean that you restart the full process from scratch but not a random init.

The above behaviour would be what I expect if I do a `partial_fit` because in this case, I am starting from a trained model and incrementally learning.

Reply:
My expectation came from `GradientBoostingClassifier`'s `warm_start`. It doesn't have a `partial_fit()`, so calling `fit()` is the only game in town.

Reply:
So maybe at the very least there is room for improving the consistency?

Reply:
We should just make it clear what to expect. But we can delay that in the issue that I open: https://github.com/scikit-learn/scikit-learn/issues/24764